appserver/distributions/glassfish-common/src/main/resources/bin/asadmin.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2025 Contributors to the Eclipse Foundation.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package org.glassfish.main.productroot.bin.asadmin;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Supplier;
+
+public class asadmin {
+
+    private static final String SYS_LAUNCHERFILE = "jdk.launcher.sourcefile";
+
+    private static final boolean AS_TRACE = "true".equals(System.getenv("AS_TRACE"));
+    private static final boolean INHERIT_ENV = "false".equals(System.getenv("AS_INHERIT_ENVIRONMENT"));
+    private static final boolean USE_SCRIPT = "true".equals(System.getenv("AS_USE_NATIVE_SCRIPT"));
+    private static final boolean IS_WINDOWS = System.getProperty("os.name").toLowerCase().contains("windows");
+
+    private static final String JAVA_HOME = System.getProperty("java.home");
+    private static final Path MY_JAVA_FILE = getMyJavaFile();
```

Comment:
The `NullPointerException` is thrown when `SYS_LAUNCHER` property is not set (see L:95).

Reply:
How? There is System.exit call if it is null

Reply:
Fixed, I made some refactoring to make it more visible next time.